source/common/common/shared_hash_map.h

```diff
@@ -0,0 +1,373 @@
+   * Note that the map dtaa itself is not constructed when the object is constructed.
+   * After the control-structure is constructed, the number of bytes can be computed so
+   * that a shared-memory segment can be allocated and passed to init() or attach().
+   */
+  SharedHashMap(const SharedHashMapOptions& options)
+      : options_(options), cells_(nullptr), control_(nullptr), slots_(nullptr) {}
+
+  /**
+   * Returns the numbers of byte required for the hash-table, based on
+   * the control structure. This must be used to allocate the
+   * backing-store (eg) in shared memory, which we do after
+   * constructing the object with the desired sizing.
+   */
+  size_t numBytes() const {
+    size_t size =
+        cellOffset(options_.capacity) + sizeof(Control) + options_.num_slots * sizeof(uint32_t);
+    return align(size);
+  }
+
+  /**
+   * Attempts to attach to an existing shared memory segment. Does a (relatively) quick
+   * sanity check to make sure the options copied to the provided memory match, and also
+   * that the slot, cell, and key-string structures look sane.
+   *
+   * Note that if mutex is in a locked state at the time of attachment, this function
+   * can hang.
+   */
+  bool attach(uint8_t* memory) {
+    mapMemorySegments(memory);
+    return sanityCheck();
+  }
+
+  /** Locks the map and runs sanity checks */
+  bool sanityCheck() LOCKS_EXCLUDED(control_->mutex) {
+    lock(); // might hang if program previously crashed.
+    bool ret = sanityCheckLockHeld();
+    unlock();
+    return ret;
+  }
+
+  /**
+   * Returns a string describing the contents of the map, including the control
+   * bits and the keys in each slot.
+   */
+  std::string toString() {
+    std::string ret;
+    lock();
+    ret = fmt::format("options={}\ncontrol={}\n", options_.toString(), control_->toString());
+    for (uint32_t i = 0; i < options_.num_slots; ++i) {
+      ret += fmt::format("slot {}:", i);
+      for (uint32_t j = slots_[i]; j != Sentinal; j = getCell(j).next_cell) {
+        ret += " " + std::string(getCell(j).getKey());
+      }
+      ret += "\n";
+    }
+    unlock();
+    return ret;
+  }
+
+  /**
+   * Initializes a hash-map on raw memory. No expectations are made about the state of the memory
+   * coming in.
+   * @param memory
+   */
+  void init(uint8_t* memory) {
+    mapMemorySegments(memory);
+    lock();
+
+    control_->options = options_;
+    control_->size = 0;
+    control_->free_cell_index = 0;
+
+    // Initialize all the slots;
+    for (uint32_t slot = 0; slot < options_.num_slots; ++slot) {
+      slots_[slot] = Sentinal;
+    }
+
+    // Initialize all the key-char offsets.
+    for (uint32_t cell_index = 0; cell_index < options_.capacity; ++cell_index) {
+      Cell& cell = getCell(cell_index);
+      cell.key_size = 0;
+      cell.next_cell = cell_index + 1;
+    }
+
+    unlock();
+  }
+
+  /**
+   * Puts a new key into the map. If successful (e.g. map has capacity)
```

Comment:
Please document that this returns existing.

Reply:
On second thought, I think the interface needs to indicate whether it added a new one or returned existing.  Or you need to run a placement new on Value in the case that it is new.  But somehow we have to be able to initialize a new one, and not initialize (ie overwrite) an existing one.

Reply:
Good point.  Done.  Returned a pair, a la STL map insert.